### ISSUE TITLE
Acceleration for (Directed)GraphSAGELinkGenerator

### ISSUE THREAD

**Issue by guomk** [stellargraph/stellargraph#1590]
### Describe the bug

I have a directed graph of about 2e4 nodes and 5e6 edges. And I was trying to feed the `UnsupervisedSampler` to `DirectedGraphSAGELinkGenerator` but the process takes too long to finish. I would be nicer if this process could be accelerated and there should be some log to help track the process (Currently there is no way to tell how many samplings have been performed.)

### To Reproduce
I couldn't disclose the dataset but any large dataset should be able to reproduce this behavior. Specifically, this behavior happens when executing this code block.

```python
biased_walker = sg.data.BiasedRandomWalk(G, n=number_of_walks, length=length, p=2, q=.5)
unsupervised_samples = UnsupervisedSampler(G, nodes=nodes, walker=biased_walker)

generator = DirectedGraphSAGELinkGenerator(G, batch_size, in_samples=in_samples, out_samples=out_samples)
train_gen = generator.flow(unsupervised_samples)
```

### Observed behavior

The whole code block takes several hours to finish. Python kernel is shown as busy. No log output.
90+% GPU memory usage

Jupyter Lab log shows
> W tensorflow/core/kernels/data/generator_dataset_op.cc:103] Error occurred when finalizing GeneratorDataset iterator: Cancelled: Operation was cancelled

### Expected behavior

Faster execution with log.

### Environment

OS: Windows 10

Python version:  *3.7.7*

Package versions: *stellargraph==1.0.0, tensorflow==2.1.0*

Code running on RTX2070 Super (8GB)

### Additional context

N/A

**Comment by kieranricardo** [comment]
Hi, thanks for filing an issue!

To help us reproduce your issues could you share your model paramaters. What is your `batch_size`, `in_samples`, `out_samples`, `number_of_walks`, and how many nodes are in `nodes`?

You also mention that your GPU usage is at 90+%, are you refferring to `Volatile GPU-Util` or `GPU Memory Usage`?

**Comment by guomk** [comment]
Hi, Thanks for your reply.

Parameter setting
`batch_size=50`
`in_samples=[10, 5]`
`out_samples=[10, 5]`
`number_of_walks=32`
`len(nodes)=16555`

The GPU usage I was referring to is memory usage.

I understand it's a large graph and the sampling is supposed to take a while to finish. But I'm just wondering whether it's utilizing the GPU well and should there be a log to show how much has finished. Could you also enlighten me on what's the complexity of the sampling algorithm regarding `batch_size`, `num_samples`, and `number_of_walks`?

**Comment by huonw** [comment]
It's not directly answering the question (I'll let @kieranricardo talk to that), but if you're interested in training an algorithm in an unsupervised fashion, another approach might be `DeepGraphInfomax` ([docs](https://stellargraph.readthedocs.io/en/latest/api.html#stellargraph.layer.DeepGraphInfomax), [demo](https://stellargraph.readthedocs.io/en/stable/demos/embeddings/deep-graph-infomax-embeddings.html)). For instance: https://nbviewer.jupyter.org/gist/huonw/aabad84802783927c9bdd6190b04f713

**Comment by kieranricardo** [comment]
> The GPU usage I was referring to is memory usage. However, I just tried to reproduce the pipeline and this time GPU memory usage is very low and utilization is about 0-1%

Ah that make sense, the code you shared doesn't use the GPU. GPU memory use is a slightly deceiving metric because when tensorflow starts a GPU process it allocates almost all of our GPU memory even if it isn't currently using it. `Volatile GPU-Util` indicates how much of you GPU is currently being used for calculations.

> Could you also enlighten me on what's the complexity of the sampling algorithm with respect tobatch_size, num_samples, and number_of_walks?

The sampling algorithm performs random walks to get pairs of target nodes, this the bit that you're stuck on! The computational complexity is `O(number_of_nodes * number_of_walks * length)`. 

I was able to replicate your issue replaced some python code with more optimized numpy code, speeding things up by about 10x on a random graph the same size as yous: https://github.com/stellargraph/stellargraph/pull/1596. Hopefully we'll merge this soon and speed things up a bit!

Also, to avoid repeating the sampling every epoch you can set `shuffle=False`:

```
train_gen = generator.flow(unsupervised_samples, shuffle=False)
```

**Comment by kieranricardo** [comment]
To follow up on @huonw's response, `DeepGraphInfomax` avoids the initial sampling step that you're stuck on and might give you better results 💯

**Comment by guomk** [comment]
Appreciate the amendments and advice!

**Comment by guomk** [comment]
I see you have made great improvements and initiated a PR! Can I check when can I expected the improvement to be merged into the master branch so that I can update my version correspondingly?

**Comment by kieranricardo** [comment]
^ just merged the change into develop, and ended up speeding things up by 25x! 

We're planning another release sometime next week so you'll have access to the speedup in master then. But if you want to try it out in the meantime you can install `stellargraph` from develop: https://github.com/stellargraph/stellargraph#install-stellargraph-from-github-source

**Comment by guomk** [comment]
Wonderfulll! Thanks so much!
